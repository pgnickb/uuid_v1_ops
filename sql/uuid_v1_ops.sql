\unset ECHO

/* setup */
\i test_setup.sql

/* try and create extension */
create extension uuid_v1_ops;

/* seed random for consistent dataset */
select setseed(0.2);

select plan(36);
/* tests: */

/* comparison */

select is(
            uuid_v1_cmp(uuid'00000000-0000-0000-0000-000000000000',
                        null::uuid
                        ), 
            null, 
            'uuid_v1_cmp handles nulls properly');

select is(
            uuid_v1_cmp(uuid'00000000-0000-0000-0000-000000000000',
                        uuid'bc2cd750-a63b-11ed-84a2-123456789112'
                        ) < 0, 
            true, 
            'Less then works fine');
select is(
            uuid_v1_cmp(uuid'bc2cd750-a63b-11ed-84a2-123456789112', 
                        uuid'00000000-0000-0000-0000-000000000000'
                        ) > 0, 
            true, 
            'Greater then works fine');
select is(
            uuid_v1_cmp(uuid'00000000-0000-0000-0000-000000000000', 
                        uuid'00000000-0000-0000-0000-000000000000'
                      ) = 0, 
            true, 
            'Equals works fine');

/* get_timesamptz tests*/

select is(
            uuid_v1_get_timestamptz(uuid'00000000-0000-0000-0000-000000000000') at time zone 'UTC', 
            timestamp'1582-10-15 00:00:00', 
            'Zero-uuid is the start of Gregorian calendar');

select is(
            uuid_v1_get_timestamptz(uuid'63b00000-bfde-11d3-0000-000000000000') at time zone 'UTC', 
            timestamp'2000-01-01 00:00:00', 
            'Timestamp of PostgreSQL Epoch');
select is(
            uuid_v1_get_timestamptz(uuid'bc2cd750-a63b-11ed-84a2-123456789112') at time zone 'UTC', 
            timestamp'2023-02-06 16:31:40.869', 
            'Some random timestamp');

/* get_timesamptz tests*/

select is(
            uuid_v1_get_timestamp_as_int8(uuid'00000000-0000-0000-0000-000000000000'), 
            0::int8, 
            'Zero-uuid is 0 int8');

select is(
            uuid_v1_get_timestamp_as_int8(uuid'63b00000-bfde-11d3-0000-000000000000'), 
            131659776000000000::int8, 
            'PostgreSQL Epoch is a proper int');
select is(
            uuid_v1_get_timestamp_as_int8(uuid'bc2cd750-a63b-11ed-84a2-123456789112'), 
            138949939008690000::int8,
            'Some random timestamp is a proper int8');

/* is_uuid works fine */

select is(
            is_uuid_v1(uuid'00000000-0000-1000-0000-000000000000'),
            true,
            'This UUID is correctly recognized as v1');

select is(
            is_uuid_v1(uuid'00000000-0000-2000-0000-000000000000'),
            false,
            'This UUID is correctly recognized as non-v1');

select is(
            is_uuid_v1(null::uuid),
            null,
            'This UUID is correctly recognized as non-v1');

/* tests with various operators and their interactions with nulls */

select is(uuid'00000000-0000-2000-0000-000000000000' ~<  null, null,  '~< handles null properly');
select is(uuid'00000000-0000-2000-0000-000000000000' ~<= null, null, '~<= handles null properly');
select is(uuid'00000000-0000-2000-0000-000000000000' =   null, null,   '= handles null properly');
select is(uuid'00000000-0000-2000-0000-000000000000' ~>= null, null, '~>= handles null properly');
select is(uuid'00000000-0000-2000-0000-000000000000' ~>  null, null,  '~> handles null properly');

/* get_node_id tests */
select is(uuid_v1_get_node_id('00000000-0000-1000-b000-000000000000'), '00:00:00:00:00:00',  'All-zeroes node id address');
select is(uuid_v1_get_node_id('00000000-0000-1000-b000-ffffffffffff'), 'ff:ff:ff:ff:ff:ff',  'All-ones node id address');
select is(uuid_v1_get_node_id('00000000-0000-0000-0000-123456789abc'), '12:34:56:78:9a:bc',  'Some random node id address');
select is(uuid_v1_get_node_id(null), null,  'get_node_id handles nulls properly');

/* get_clock_seq tests */
select is(uuid_v1_get_clock_seq('00000000-0000-1000-8000-000000000000'), 0::int2,  'All-zeroes clock-sequence');
select is(uuid_v1_get_clock_seq('00000000-0000-1000-7fff-000000000000'), 16383::int2,  'All-ones clock-sequence');
select is(uuid_v1_get_clock_seq('00000000-0000-1000-9357-000000000000'), 4951::int2,  'Some random clock-sequence');
select is(uuid_v1_get_clock_seq(null), null,  'get_clock_seq handles nulls properly');

/* get_variant tests */
select is(uuid_v1_get_variant('00000000-0000-1000-0000-000000000000'), 0::int2,  'All-zeroes variant');
select is(uuid_v1_get_variant('00000000-0000-1000-4000-000000000000'), 1::int2,  'Variant 1');
select is(uuid_v1_get_variant('00000000-0000-1000-8000-000000000000'), 2::int2,  'Variant 2');
select is(uuid_v1_get_variant('00000000-0000-1000-c000-000000000000'), 3::int2,  'Variant 3');
select is(uuid_v1_get_variant(null), null,  'get_variant handles nulls properly');

/* create_from_ts tests */

select is(
        uuid_v1_create_from_ts(now(), null, null),
        null,
        'create_from_ts handles nulls properly');

/* xxx in principle this is a strict test, as precision of 
 * the PostgreSQL timestamptz type is lower that that of the
 * UUID v1 (1usec vs 0.1usec respectively).
 * Meaning if the UUID was created using this function,
 * it is inevitably going to be consistently reversible.
 */

with gen as (
    select
            u,
            uuid_v1_create_from_ts(
                        uuid_v1_get_timestamptz(u),
                        uuid_v1_get_clock_seq(u),
                        uuid_v1_get_node_id(u)) as g
        from generate_series('2020-10-20 00:00:00 UTC',
                             '2020-10-30 00:00:00 UTC',
                             interval'20 minutes') as f(x),
        uuid_v1_create_from_ts(x,
                            0::int2,
                            macaddr'12:34:56:78:91:12') as g(u)
), cnt as(
    select count(*) as c
        from gen
        where u != g
)
select is(c, 0::int8, 'Timestamptz UUID generation is consistent') from cnt;

/* create_from_int8 tests */

select is(
        uuid_v1_create_from_int8(
                uuid_v1_get_timestamp_as_int8(u),
                uuid_v1_get_clock_seq(u),
                uuid_v1_get_node_id(u)),
        u,
        'int8 UUIDs generation is consistent')
    from (values (uuid'c7485dbc-ae15-11ed-a708-012345678901'),
                 (uuid'c7e73842-ae15-11ed-be05-012345678901'))
         as f(u);

select is(
        uuid_v1_create_from_int8(null, null, null),
        null,
        'create_from_int8 handles nulls properly');

select * from finish();
