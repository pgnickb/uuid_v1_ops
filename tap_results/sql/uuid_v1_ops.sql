1..6
ok 1 - Less then works fine
ok 2 - Greater then works fine
ok 3 - Equals works fine
ok 4 - Zero-uuid is the start of Gregorian calendar
ok 5 - Timestamp of PostgreSQL Epoch
ok 6 - Some random timestamp
