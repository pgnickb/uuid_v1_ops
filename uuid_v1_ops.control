# uuid_v1_ops extension
comment = 'uuid_v1_ops - adds support for UUID v1 specific btree operator class'
default_version = '0.1.0'
relocatable = true
module_pathname = 'ibdir/uuid_v1_ops'
