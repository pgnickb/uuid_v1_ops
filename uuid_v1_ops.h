#ifndef UUID_V1_OPS_H
#define UUID_V1_OPS_H

#include "postgres.h"
#include "nodes/print.h"
#include "utils/builtins.h"
#include "utils/sortsupport.h"
#include "utils/uuid.h"
#include "utils/inet.h"
#include "utils/timestamp.h"

#define UUID_V1_100NS_TO_USEC INT64CONST(10)
#define UUID_V1_TIMESTAMP_LEN 8
#define GREGORIAN_BEGINNING_OFFSET_USEC INT64CONST(13165977600000000)
#define UUID_VERSION_OFFSET 6

#define UUID_V1_GREATEST_SUPPORTED_TIMESTAMP INT64CONST(909171226085477580)
#define UUID_V1_GREATEST_SUPPORTED_INT8 INT64CONST(0xf000000000000000)
#define UUID_V1_LEAST_SUPPORTED_TIMESTAMP -GREGORIAN_BEGINNING_OFFSET_USEC

#define UUID_V1_NODE_OFFSET_A 10
#define UUID_V1_NODE_OFFSET_B 11
#define UUID_V1_NODE_OFFSET_C 12
#define UUID_V1_NODE_OFFSET_D 13
#define UUID_V1_NODE_OFFSET_E 14
#define UUID_V1_NODE_OFFSET_F 15

#define UUID_V1_SEQ_OFFSET 8

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(is_uuid_v1);
Datum		is_uuid_v1(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_cmp);
Datum		uuid_v1_cmp(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_timestamptz);
Datum		uuid_v1_get_timestamptz(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_timestamp_as_int8);
Datum		uuid_v1_get_timestamp_as_int8(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_lt);
Datum		uuid_v1_lt(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_le);
Datum		uuid_v1_le(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_ge);
Datum		uuid_v1_ge(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_gt);
Datum		uuid_v1_gt(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_node_id);
Datum		uuid_v1_get_node_id(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_clock_seq);
Datum		uuid_v1_get_clock_seq(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_get_variant);
Datum		uuid_v1_get_variant(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_create_from_ts);
Datum		uuid_v1_create_from_ts(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(uuid_v1_create_from_int8);
Datum		uuid_v1_create_from_int8(PG_FUNCTION_ARGS);

pg_uuid_t       *uuid_v1_create_from_internal(int64 ts, int16 clock_seq, macaddr *node);
static int64    uuid_v1_get_timestamp_internal(pg_uuid_t *uuid);
static int      uuid_v1_internal_cmp(const pg_uuid_t *arg1, const pg_uuid_t *arg2);

static unsigned int order[UUID_LEN] = {6, 7, 4, 5, 0, 1, 2, 3, 8, 9, 10, 11, 12, 13, 14, 15};

#endif							/* UUID_V1_OPS_H */
