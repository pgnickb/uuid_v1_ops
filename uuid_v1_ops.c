#include <uuid_v1_ops.h>

Datum
is_uuid_v1(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);

	PG_RETURN_BOOL(
				   (arg1->data[UUID_VERSION_OFFSET] & 0xf0) == 0x10
		);
}

Datum
uuid_v1_get_timestamptz(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	uint64		res = 0;
	uint64		gns;			/* number of 100ns intervals since EPOCH */

	gns = ((uint64) ((arg1->data[order[0]]) & 0x0F)) << 56;
	for (int i = 1; i < UUID_V1_TIMESTAMP_LEN; ++i)
	{
		/* this conversion is safe regardless of the endiannes */
		gns += ((uint64) arg1->data[order[i]]) << ((UUID_V1_TIMESTAMP_LEN - i - 1) * 8);
	}

	res = (gns / UUID_V1_100NS_TO_USEC) - GREGORIAN_BEGINNING_OFFSET_USEC;
	PG_RETURN_TIMESTAMPTZ(res);
}

Datum
uuid_v1_cmp(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_INT32(uuid_v1_internal_cmp(arg1, arg2));
}

static int
uuid_v1_internal_cmp(const pg_uuid_t *arg1, const pg_uuid_t *arg2)
{
	int			res;

	for (int i = 0; i < UUID_LEN; ++i)
	{
		res = arg1->data[order[i]] - arg2->data[order[i]];
		if (res != 0)
			break;
	}

	return res;
}

Datum
uuid_v1_lt(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) < 0);
}

Datum
uuid_v1_le(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) <= 0);
}

Datum
uuid_v1_eq(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) == 0);
}

Datum
uuid_v1_ge(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) >= 0);
}

Datum
uuid_v1_gt(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	pg_uuid_t  *arg2 = PG_GETARG_UUID_P(1);

	PG_RETURN_BOOL(uuid_v1_internal_cmp(arg1, arg2) > 0);
}

Datum
uuid_v1_get_node_id(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	macaddr    *result;

	result = (macaddr *) palloc(sizeof(macaddr));

	result->a = arg1->data[UUID_V1_NODE_OFFSET_A];
	result->b = arg1->data[UUID_V1_NODE_OFFSET_B];
	result->c = arg1->data[UUID_V1_NODE_OFFSET_C];
	result->d = arg1->data[UUID_V1_NODE_OFFSET_D];
	result->e = arg1->data[UUID_V1_NODE_OFFSET_E];
	result->f = arg1->data[UUID_V1_NODE_OFFSET_F];

	PG_RETURN_MACADDR_P(result);
}

Datum
uuid_v1_get_clock_seq(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	int16		seq;

	seq = ((int16) ((arg1->data[UUID_V1_SEQ_OFFSET]) & 0x3F)) << 8;
	seq += arg1->data[UUID_V1_SEQ_OFFSET + 1];

	PG_RETURN_INT16(seq);
}

Datum
uuid_v1_get_variant(PG_FUNCTION_ARGS)
{
	pg_uuid_t  *arg1 = PG_GETARG_UUID_P(0);
	int16		v;

	v = ((int16) ((arg1->data[UUID_V1_SEQ_OFFSET]) & 0xC0)) >> 6;

	PG_RETURN_INT16(v);
}


Datum
uuid_v1_create_from(PG_FUNCTION_ARGS)
{
	TimestampTz ts = PG_GETARG_TIMESTAMPTZ(0);
	int16		clock_seq = PG_GETARG_INT16(1);
	macaddr    *node = PG_GETARG_MACADDR_P(2);
	pg_uuid_t  *res;

	res = (pg_uuid_t *) palloc(sizeof(*res));

	/*
	 * Convert PostgreSQL epoch usec timestamptz to the UUID v1 Gregorian
	 * epoch 0.1usec This can overflow, so we need to be careful. UUID v1
	 * Timestamp is unsigned and thus can't represent dates before it's epoch.
	 */

	if (ts > UUID_V1_GREATEST_SUPPORTED_TIMESTAMP ||
		ts < UUID_V1_LEAST_SUPPORTED_TIMESTAMP)
	{
		ereport(ERROR,
				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
				 errmsg("timestamp value out of range for UUID v1")));
	}

	ts = (ts + GREGORIAN_BEGINNING_OFFSET_USEC) * UUID_V1_100NS_TO_USEC;

	/* We try to keep further modifications endiannes-neutral */

	/* timestamp low bytes */
	res->data[0] = (uint8) ((ts & 0xff000000) >> 24);
	res->data[1] = (uint8) ((ts & 0xff0000) >> 16);
	res->data[2] = (uint8) ((ts & 0xff00) >> 8);
	res->data[3] = (uint8) ((ts & 0xff));

	/* timestamp mid bytes */
	res->data[4] = (uint8) ((ts & 0xff0000000000) >> 40);
	res->data[5] = (uint8) ((ts & 0xff00000000) >> 32);

	/* timestamp hi bytes and version */
	res->data[6] = (uint8) ((ts & 0x0f00000000000000) >> 56);
	/* append version to the upper MSB of timestamp hi */
	res->data[6] |= (1 << 4);
	res->data[7] = (uint8) ((ts & 0xff000000000000) >> 48);

	/* clock seq hi and reserved */
	res->data[8] = (uint8) ((clock_seq & 0x3f00) >> 8);
	/* append reserver to the clock seq high */
	res->data[8] |= 0x80;
	/* clock seq low */
	res->data[9] = (uint8) (clock_seq & 0xff);

	/* node id */
	res->data[10] = node->a;
	res->data[11] = node->b;
	res->data[12] = node->c;
	res->data[13] = node->d;
	res->data[14] = node->e;
	res->data[15] = node->f;

	PG_RETURN_UUID_P(res);
}
