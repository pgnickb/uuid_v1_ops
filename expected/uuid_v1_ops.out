\unset ECHO

1..36
ok 1 - uuid_v1_cmp handles nulls properly
ok 2 - Less then works fine
ok 3 - Greater then works fine
ok 4 - Equals works fine
ok 5 - Zero-uuid is the start of Gregorian calendar
ok 6 - Timestamp of PostgreSQL Epoch
ok 7 - Some random timestamp
ok 8 - Zero-uuid is 0 int8
ok 9 - PostgreSQL Epoch is a proper int
ok 10 - Some random timestamp is a proper int8
ok 11 - This UUID is correctly recognized as v1
ok 12 - This UUID is correctly recognized as non-v1
ok 13 - This UUID is correctly recognized as non-v1
ok 14 - ~< handles null properly
ok 15 - ~<= handles null properly
ok 16 - = handles null properly
ok 17 - ~>= handles null properly
ok 18 - ~> handles null properly
ok 19 - All-zeroes node id address
ok 20 - All-ones node id address
ok 21 - Some random node id address
ok 22 - get_node_id handles nulls properly
ok 23 - All-zeroes clock-sequence
ok 24 - All-ones clock-sequence
ok 25 - Some random clock-sequence
ok 26 - get_clock_seq handles nulls properly
ok 27 - All-zeroes variant
ok 28 - Variant 1
ok 29 - Variant 2
ok 30 - Variant 3
ok 31 - get_variant handles nulls properly
ok 32 - create_from_ts handles nulls properly
ok 33 - Timestamptz UUID generation is consistent
ok 34 - int8 UUIDs generation is consistent
ok 35 - int8 UUIDs generation is consistent
ok 36 - create_from_int8 handles nulls properly
